include/hydrosheds/hydrosheds.hh

```diff
@@ -4,59 +4,193 @@
 
 #include<array>
 #include<string>
+#include <cmath>
 
 namespace hydrosheds {
 
-  using Coordinate = std::array<double, 2>;
-  using HydroshedsID = int;
+    using Coordinate = std::array<double, 2>;
+    using HydroshedsID = int;
 
-  // Forward declarations
-  class RiverSegment;
-  class FullDatasetRiverSegmentIterator;
+    // Forward declarations
+    class RiverSegment;
+    class FullDatasetRiverSegmentIterator;
 
-  class HydroshedsDataSet
-  {
+    class DownstreamIterator;
+    class DownstreamIteratorHelper;
+
+
+    class HydroshedsDataSet
+    {
     public:
-    HydroshedsDataSet(const std::string&);
+        HydroshedsDataSet(const std::string&);
+
+        // Iterator access for Feature Iteration
 
-    // Iterator access
-    FullDatasetRiverSegmentIterator begin() const;
-    FullDatasetRiverSegmentIterator end() const;
+        FullDatasetRiverSegmentIterator begin() const;
+        FullDatasetRiverSegmentIterator end() const;
 
-    // Not intended to stay - just get some segment
-    RiverSegment getSegment() const;
+        // Iterator access for Segment Iteration
+        DownstreamIteratorHelper followbegin();
```

Comment:
I think it is safe marking these methods `const`, just like in the `FullDatasetRiverSegmentIterator` case.